<?xml version="1.0" encoding="UTF-8"?>
<robot name="so101_follower">
  <joint name="elbow_flex" type="revolute">
    <origin xyz="-0.11257 -0.028 -0. " rpy="-0. -0. 1.5708"/>
    <parent link="upper_arm"/>
    <child link="lower_arm"/>
    <axis xyz="0. 0. 1."/>
    <limit lower="-1.7453299" upper="1.5707999" effort="10." velocity="9.9999995"/>
  </joint>
  <joint name="joints_gripper" type="revolute">
    <origin xyz="0.0202 0.0188 -0.0234" rpy="1.5708001 0.0000002 0.0000005"/>
    <parent link="so101_new_calib_gripper"/>
    <child link="jaw"/>
    <axis xyz="0. 0. 1."/>
    <limit lower="-0.174533" upper="1.7453299" effort="10." velocity="9.9999995"/>
  </joint>
  <joint name="shoulder_lift" type="revolute">
    <origin xyz="-0.0303992 -0.0182778 -0.0542 " rpy="1.5544026 -1.5707927 -3.1252024"/>
    <parent link="shoulder"/>
    <child link="upper_arm"/>
    <axis xyz="0. 0. 1."/>
    <limit lower="-1.7453299" upper="1.7453299" effort="10." velocity="9.9999995"/>
  </joint>
  <joint name="shoulder_pan" type="revolute">
    <origin xyz="0.0207909 -0.0230745 0.0948817" rpy="-3.14159 0. 1.5708 "/>
    <parent link="so101_new_calib"/>
    <child link="shoulder"/>
    <axis xyz="0. 0. 1."/>
    <limit lower="-1.9198599" upper="1.9198599" effort="10." velocity="9.9999995"/>
  </joint>
  <joint name="wrist_flex" type="revolute">
    <origin xyz="-0.1349 0.0052 -0. " rpy="0.0000001 0. -1.5708001"/>
    <parent link="lower_arm"/>
    <child link="wrist"/>
    <axis xyz="0. 0. 1."/>
    <limit lower="-1.6580599" upper="1.6580599" effort="10." velocity="9.9999995"/>
  </joint>
  <joint name="wrist_roll" type="revolute">
    <origin xyz="0. -0.0611 0.0181" rpy="1.5708001 0.0000001 3.14159 "/>
    <parent link="wrist"/>
    <child link="so101_new_calib_gripper"/>
    <axis xyz="0. 0. 1."/>
    <limit lower="-2.79253" upper="2.79253" effort="10." velocity="9.9999995"/>
  </joint>
  <link name="jaw">
    <inertial>
      <origin xyz="-0.0015749 -0.0300244 0.0192755" rpy="0. 0. 0."/>
      <mass value="0.012"/>
      <inertia ixx="0.0000066" ixy="-0.0000003" ixz="-0." iyy="0.0000019" iyz="-0.0000001" izz="0.0000053"/>
    </inertial>
    <visual>
      <origin xyz="0. -0. 0.0189" rpy="-0.0000001 0.0000004 0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/jaw_visuals_moving_jaw_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0. -0. 0.0189" rpy="-0.0000001 0.0000004 0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/jaw_collisions_moving_jaw_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0. -0. 0.0189" rpy="-0.0000001 0.0000004 0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/jaw_collisions_moving_jaw_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
  <link name="lower_arm">
    <inertial>
      <origin xyz="-0.0980701 0.0032438 0.0182831" rpy="0. 0. 0."/>
      <mass value="0.104"/>
      <inertia ixx="0.0000287" ixy="0.0000074" ixz="0.0000013" iyy="0.0001598" iyz="-0." izz="0.0001453"/>
    </inertial>
    <visual>
      <origin xyz="-0.0648499 -0.032 0.0182 " rpy="-3.14159 -0.0000001 -0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_visuals_under_arm_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.0648499 -0.032 0.018 " rpy="-3.14159 -0.0000001 -0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_visuals_motor_holder_so101_wrist_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.1224 0.0052 0.0187" rpy="-3.14159 0.0000001 -3.1415904"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_visuals_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.0648499 -0.032 0.0182 " rpy="-3.14159 -0.0000001 -0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_collisions_under_arm_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.0648499 -0.032 0.018 " rpy="-3.14159 -0.0000001 -0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_collisions_motor_holder_so101_wrist_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.1224 0.0052 0.0187" rpy="-3.14159 0.0000001 -3.1415904"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0648499 -0.032 0.0182 " rpy="-3.14159 -0.0000001 -0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_collisions_under_arm_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.0648499 -0.032 0.018 " rpy="-3.14159 -0.0000001 -0.0000001"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_collisions_motor_holder_so101_wrist_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.1224 0.0052 0.0187" rpy="-3.14159 0.0000001 -3.1415904"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/lower_arm_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
  <link name="shoulder">
    <inertial>
      <origin xyz="-0.0307604 -0.0000167 -0.0252713" rpy="0. 0. 0."/>
      <mass value="0.100006"/>
      <inertia ixx="0.0000838" ixy="0.0000001" ixz="-0.0000012" iyy="0.000081" iyz="0.0000002" izz="0.000024"/>
    </inertial>
    <visual>
      <origin xyz="-0.0303992 0.0004222 -0.0417 " rpy="-1.5488194 1.5707927 -3.1196194"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_visuals_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.0675992 -0.0001778 0.0158499" rpy="-1.5488194 -1.5707927 3.1196194"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_visuals_motor_holder_so101_base_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0122008 0.0000222 0.0464 " rpy="-1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_visuals_rotation_pitch_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.0303992 0.0004222 -0.0417 " rpy="-1.5488194 1.5707927 -3.1196194"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.0675992 -0.0001778 0.0158499" rpy="-1.5488194 -1.5707927 3.1196194"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_collisions_motor_holder_so101_base_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0122008 0.0000222 0.0464 " rpy="-1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_collisions_rotation_pitch_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.0303992 0.0004222 -0.0417 " rpy="-1.5488194 1.5707927 -3.1196194"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.0675992 -0.0001778 0.0158499" rpy="-1.5488194 -1.5707927 3.1196194"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_collisions_motor_holder_so101_base_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.0122008 0.0000222 0.0464 " rpy="-1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/shoulder_collisions_rotation_pitch_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
  <link name="so101_new_calib">
    <visual>
      <origin xyz="0.0206915 0.0221255 0.0300817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_visuals_base_motor_holder_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0207909 0.0221255 0.0300817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_visuals_base_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0207909 -0.0105745 0.0761817" rpy="0. 0. -1.5708"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_visuals_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0205915 0.0467435 0.0798817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_visuals_waveshare_mounting_plate_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0206915 0.0221255 0.0300817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_base_motor_holder_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0207909 0.0221255 0.0300817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_base_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0207909 -0.0105745 0.0761817" rpy="0. 0. -1.5708"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0205915 0.0467435 0.0798817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_waveshare_mounting_plate_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.0206915 0.0221255 0.0300817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_base_motor_holder_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.0207909 0.0221255 0.0300817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_base_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.0207909 -0.0105745 0.0761817" rpy="0. 0. -1.5708"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.0205915 0.0467435 0.0798817" rpy="1.5708 0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/base_collisions_waveshare_mounting_plate_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
  <link name="so101_new_calib_gripper">
    <inertial>
      <origin xyz="0.0002136 0.0002451 -0.025187 " rpy="0. 0. 0."/>
      <mass value="0.087"/>
      <inertia ixx="0.0000275" ixy="-0.0000003" ixz="-0.0000057" iyy="0.0000434" iyz="-0.0000001" izz="0.0000345"/>
    </inertial>
    <visual>
      <origin xyz="0.0077 0.0001 -0.0234" rpy="-1.5707998 0.0000002 0.0000005"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/so101_new_calib_gripper_visuals_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0. -0.0002182 0.0009497" rpy="-3.14159 0.0000002 0.0000005"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/so101_new_calib_gripper_visuals_wrist_roll_follower_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0.0077 0.0001 -0.0234" rpy="-1.5707998 0.0000002 0.0000005"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/so101_new_calib_gripper_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0. -0.0002182 0.0009497" rpy="-3.14159 0.0000002 0.0000005"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/so101_new_calib_gripper_collisions_wrist_roll_follower_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.0077 0.0001 -0.0234" rpy="-1.5707998 0.0000002 0.0000005"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/so101_new_calib_gripper_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0. -0.0002182 0.0009497" rpy="-3.14159 0.0000002 0.0000005"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/so101_new_calib_gripper_collisions_wrist_roll_follower_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
  <link name="upper_arm">
    <inertial>
      <origin xyz="-0.0898471 -0.0083822 0.0184089" rpy="0. 0. 0."/>
      <mass value="0.103"/>
      <inertia ixx="0.0000408" ixy="-0.0000198" ixz="-0." iyy="0.0001473" iyz="0." izz="0.0001425"/>
    </inertial>
    <visual>
      <origin xyz="-0.11257 -0.0155 0.0187 " rpy="-3.14159 0. -1.5708 "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/upper_arm_visuals_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.065085 0.012 0.0182 " rpy="3.14159 -0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/upper_arm_visuals_upper_arm_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.11257 -0.0155 0.0187 " rpy="-3.14159 0. -1.5708 "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/upper_arm_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="-0.065085 0.012 0.0182 " rpy="3.14159 -0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/upper_arm_collisions_upper_arm_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.11257 -0.0155 0.0187 " rpy="-3.14159 0. -1.5708 "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/upper_arm_collisions_sts3215_03a_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="-0.065085 0.012 0.0182 " rpy="3.14159 -0. 0. "/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/upper_arm_collisions_upper_arm_so101_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
  <link name="wrist">
    <inertial>
      <origin xyz="-0.0001033 -0.0386143 0.0281156" rpy="0. 0. 0."/>
      <mass value="0.079"/>
      <inertia ixx="0.0000368" ixy="0." ixz="-0.0000001" iyy="0.0000254" iyz="0.0000036" izz="0.000021"/>
    </inertial>
    <visual>
      <origin xyz="0. -0.0424 0.0306" rpy="-1.5307784 1.5707929 -3.1015782"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/wrist_visuals_sts3215_03a_no_horn_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0. -0.028 0.0181" rpy="1.5649078 -1.5707926 -3.1357076"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/wrist_visuals_wrist_roll_pitch_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0. -0.0424 0.0306" rpy="-1.5307784 1.5707929 -3.1015782"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/wrist_collisions_sts3215_03a_no_horn_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <visual>
      <origin xyz="0. -0.028 0.0181" rpy="1.5649078 -1.5707926 -3.1357076"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/wrist_collisions_wrist_roll_pitch_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0. -0.0424 0.0306" rpy="-1.5307784 1.5707929 -3.1015782"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/wrist_collisions_sts3215_03a_no_horn_v1_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="0. -0.028 0.0181" rpy="1.5649078 -1.5707926 -3.1357076"/>
      <geometry>
        <mesh filename="file:///home/luohefei/Downloads/urdf/meshes/wrist_collisions_wrist_roll_pitch_so101_v2_mesh.obj" scale="1. 1. 1."/>
      </geometry>
    </collision>
  </link>
</robot>
